### DISCUSSION TITLE
How to stop GPU occupancy after the model end training?

### DISCUSSION THREAD

**Discussion by pzSuen** [ContinualAI/avalanche#848] [Q&A]
I had run the code in "Putting All Together," but I found that the GPU is still occupying when the training is over.
How can I stop it?

**Comment by vlomonaco** [comment]
Are you sure this is related to Avalanche and not Pytorch? Please provide the snippet of code to reproduce the issue! :)

**Reply by pzSuen** [reply]
Just code copy for "putting all together," and add a line to use the GPU.
```
from torch.optim import SGD
from torch.nn import CrossEntropyLoss
from avalanche.benchmarks.classic import SplitMNIST
from avalanche.evaluation.metrics import forgetting_metrics, accuracy_metrics, \
    loss_metrics, timing_metrics, cpu_usage_metrics, confusion_matrix_metrics, disk_usage_metrics
from avalanche.models import SimpleMLP
from avalanche.logging import InteractiveLogger, TextLogger, TensorboardLogger
from avalanche.training.plugins import EvaluationPlugin
from avalanche.training.strategies import Naive

# Using this device
device = torch.device("cuda:1" if torch.cuda.is_available() else "CPU")

scenario = SplitMNIST(n_experiences=5)

# MODEL CREATION
model = SimpleMLP(num_classes=scenario.n_classes)

# DEFINE THE EVALUATION PLUGIN and LOGGERS
# The evaluation plugin manages the metrics computation.
# It takes as argument a list of metrics, collectes their results and returns
# them to the strategy it is attached to.

# log to Tensorboard
tb_logger = TensorboardLogger()

# log to text file
text_logger = TextLogger(open('log.txt', 'a'))

# print to stdout
interactive_logger = InteractiveLogger()

eval_plugin = EvaluationPlugin(
        accuracy_metrics(minibatch=True, epoch=True, experience=True, stream=True),
        loss_metrics(minibatch=True, epoch=True, experience=True, stream=True),
        timing_metrics(epoch=True, epoch_running=True),
        forgetting_metrics(experience=True, stream=True),
        cpu_usage_metrics(experience=True),
        confusion_matrix_metrics(num_classes=scenario.n_classes, save_image=False,
                                 stream=True),
        disk_usage_metrics(minibatch=True, epoch=True, experience=True, stream=True),
        loggers=[interactive_logger, text_logger, tb_logger]
)

# CREATE THE STRATEGY INSTANCE (NAIVE)
cl_strategy = Naive(
        model, SGD(model.parameters(), lr=0.001, momentum=0.9),
        CrossEntropyLoss(), train_mb_size=500, train_epochs=1, eval_mb_size=100,
        evaluator=eval_plugin, device=device) # add device

# TRAINING LOOP
print('Starting experiment...')
results = []
for experience in scenario.train_stream:
    print("Start of experience: ", experience.current_experience)
    print("Current Classes: ", experience.classes_in_this_experience)

    # train returns a dictionary which contains all the metric values
    res = cl_strategy.train(experience)
    print('Training completed')

    print('Computing accuracy on the whole test set')
    # test also returns a dictionary which contains all the metric values
    results.append(cl_strategy.eval(scenario.test_stream))

```

And as the image shows, the program had stopped evaluation, but the GPU occupancy wasn't stopped with GPU-Util is 0.

![image](https://user-images.githubusercontent.com/43957007/145560509-77de598d-73c6-4fcf-97e4-9dda5401dac1.png)

**Reply by pzSuen** [reply]
@vlomonaco Hello, 
I solved this problem by not using TensorBoard (delete "tb_logger = TensorboardLogger()" and "tb_logger" in eval_plugin). 
I think this is because Avalanche doesn't close the TensorBoard writer after training (The code was entirely copied from Avalanche Doc with nothing neglect).

**Reply by vlomonaco** [reply]
Thanks @pzSuen! @AndreaCossu can you confirm this is a bug in the TB logging plugin?

**Reply by AndreaCossu** [reply]
Not really, I think. The loggers do not have to be closed after training, since you may call the `strategy.train`, `strategy.eval` in the future. The loggers have to be closed only when the strategy object is deleted, i.e. when the script closes, in most of the cases. 
For `TensorboardLogger`, an alternative would be to manually close the logger by `del logger`, provided that you have a reference to the logger object. Let me know :smile:

**Reply by pzSuen** [reply]
> 

Thank you very much, but I don't understand what you mean by "when the script closes", how?

**Reply by AndreaCossu** [reply]
Maybe I misunderstood your question. Are you saying that the GPU is still occupied after the `eval` call or after the entire script is finished?

**Reply by pzSuen** [reply]
I mean "after the entire script is finished".

I want a more explicit way to close the `strategy` if I don't use it later, rather than by `del logger`. Such as `strategy.close()` like `writer.close()` in Tensorboard.

**Reply by AntonioCarta** [reply]
I think @pzSuen is pointing out at an important use case. In general, since a CL system may run for a long time, we need to give the ability to stop and restart it. This entails moving models back to the cpu, closing logger files and much more. Eventually, the possibility to serialize everything to close and restart the process. 

An easy way to start to start to support this would be to close the files after every `train` and `eval` call. @AndreaCossu is this doable for the Tensorboard logger? are there any limitations?

**Reply by AndreaCossu** [reply]
The closing part is not difficult. For the resume of the training / evaluation from the point in which it has been left off, I think we need to proceed logger by logger. For Tensorboard I have to check if there is the possibility to append data to an existing run in order to have a single curve for each metric. Otherwise, we will have to create a new run which starts from the previous state (e.g. previous epoch, experience etc.).

**Reply by pzSuen** [reply]
@AntonioCarta is right. I ran a shell script to test different strategies, but it couldn't stop and continue with the GPU occupancy. 

```
#!/bin/sh
gpu='3'
for plugin in Naive Replay GDumb AGEM EWC LwF; do

  CUDA_VISIBLE_DEVICES=${gpu} python came_cls.py -a --cl_plugin ${plugin} >>out.txt

done
```

I had saved my problem by not using Tensorboard, but escaping problems is not a good solution.  
What I had said above is just advice.

Thank you~
